### PR TITLE
Popup menus RTL

### PR DESCRIPTION
**Pull request by Hixie**
This fixes the popup menu code to do a better job of expanding
smoothly regardless of which side of the screen it's on. It still
results in a bidirection growth when positioned at the bottom of the
screen, so maybe we'll need to animate menus differently, but that's
a problem for another patch.

Also, improve some docs and provide RelativeRect.toSize which I needed
at one point while building this patch (though it didn't survive all
the way to the end).

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -3130,7 +3130,7 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
   }
 }
 
-/// A [_InterstingSemanticsFragment] that can be told to only add explicit
+/// A [_InterestingSemanticsFragment] that can be told to only add explicit
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/object.dart]: I suppose this should be "An [_Interesting...]". I'm confused by the format of the entire file. Private classes with '///' comments?

File: packages/flutter/lib/src/rendering/stack.dart [hunk 2]
```diff
@@ -103,10 +103,24 @@ class RelativeRect {
   }
 
   /// Convert this [RelativeRect] to a [Rect], in the coordinate space of the container.
+  ///
+  /// See also:
+  ///
+  ///  * [toSize], which returns the size part of the rect, based on the size of
+  ///    the container.
   Rect toRect(Rect container) {
     return new Rect.fromLTRB(left, top, container.width - right, container.height - bottom);
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/stack.dart]: I'm confused by this.  On the face of it, I would have expected the return value to be:

  `new Rect.fromLTRB(left, top, left + container.width - right, top + container.height - bottom);

**Review by HansMuller — COMMENTED** [review]
LGTM

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 3]
```diff
@@ -252,6 +252,10 @@ class _ModalBottomSheetRoute<T> extends PopupRoute<T> {
 /// preventing the use from interacting with the app. Persistent bottom sheets
 /// can be created and displayed with the [ScaffoldState.showBottomSheet] function.
 ///
+/// The `context` argument is used to look up the [Navigator] and [Theme] for
+/// the bottom sheet. It is only used when the method is called, and can be
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/material/bottom_sheet.dart]: Good call. Fixed.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 4]
```diff
@@ -460,37 +460,76 @@ class _PopupMenu<T> extends StatelessWidget {
...
+    // Find the ideal horizontal position.
+    double x;
+    if (position.left > position.right) {
+      // Menu button is closer to the right edge, so grow to the left, aligned to the right edge.
+      x = size.width - position.right - childSize.width;
+    } else if (position.left < position.right) {
+      // Menu button is closer to the left edge, so grow to the right, aligned to the left edge.
+      x = position.left;
+    } else {
+      // Menu button is equidistance from both edges, so grow in reading direction.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/material/popup_menu.dart]: fixed

File: packages/flutter/lib/src/rendering/object.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/object.dart]: We use `///` for comments that are high enough quality that we could publish them if the class was public, in theory.

Comment (Hixie) [reply packages/flutter/lib/src/rendering/object.dart]: also, fixed grammar

File: packages/flutter/lib/src/rendering/stack.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/stack.dart]: RelativeRect gives the distance from the edges of the container. Rect gives the distance from the top left of the container.

**Comment by Hixie** [comment]
Will land on green unless you think I'm crazy about the toRect maths... (It's quite possible that I'm confused.)

**Comment by HansMuller** [comment]
Please land whenever you're ready. We should chat about the toRect() arithmetic.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 5]
```diff
@@ -460,37 +460,76 @@ class _PopupMenu<T> extends StatelessWidget {
...
+    // size: The size of the overlay.
+    // childSize: The size of the menu, when fully open, as determined by
+    // getConstraintsForChild.
+
+    // Find the ideal vertical position.
+    double y;
+    if (selectedItemOffset == null) {
+      y = position.top;
+    } else {
+      y = position.top + (size.height - position.top - position.bottom) / 2.0 - selectedItemOffset;
```
Comment (HubertZhang) [inline packages/flutter/lib/src/material/popup_menu.dart]: Could you explain why you need to add `(size.height - position.top - position.bottom) / 2.0` to the top offset? It makes my popup menu showing at a wrong place...

Comment (HansMuller) [reply packages/flutter/lib/src/material/popup_menu.dart]: If you're seeing a problem with popup menus, please open a new issue and include a small test case that demos the problem. Happy to take a look!

**Pull request merged**